core/Pos/Binary/Class/Core.hs

```diff
@@ -1,171 +1,613 @@
-         else throwM $ PeekException 0 "isolate64Full: size too large for 32-bit")
-#endif
-    PeekResult ptr' x <- Store.runPeek m ps ptr
-    let ptr2 = ptr `plusPtr` fromIntegral len
-    when (ptr' < ptr2) $
-        throwM $ PeekException (ptr2 `minusPtr` ptr') "Not all isolated bytes read"
-    when (ptr' > ptr2) $
-        throwM $ PeekException (ptr' `minusPtr` ptr2) "Overshot end of isolated bytes"
-    return $ PeekResult ptr2 x
-
--- | Find out how many bytes were read during a 'Peek'.
-getPeekLength :: Peek a -> Peek (a, Int)
-getPeekLength m = Peek $ \ps ptr -> do
-    PeekResult ptr2 x <- Store.runPeek m ps ptr
-    return (PeekResult ptr2 (x, ptr2 `minusPtr` ptr))
-{-# INLINE getPeekLength #-}
+    -- Decode a single (k,v).
+    decodeEntry :: (Bi k, Bi v) => D.Decoder s (k,v)
+    decodeEntry = do
+        !k <- decode
+        !v <- decode
+        return (k, v)
+
+    -- Decode all the entries, enforcing canonicity by ensuring that the
+    -- previous key is smaller or equal than the next one.
+    decodeEntries :: (Bi k, Bi v, Ord k) => Int -> k -> [(k,v)] -> D.Decoder s [(k,v)]
+    decodeEntries 0 _ acc = return (reverse acc)
+    decodeEntries !remainingPairs previousKey !acc = do
+        p@(newKey, _) <- decodeEntry
+        case newKey < previousKey of
+            True  -> fail "Canonicity violation whilst decoding a Map!"
+            False -> decodeEntries (remainingPairs - 1) newKey (p : acc)
+{-# INLINE decodeMapSkel #-}
+
+instance (Hashable k, Ord k, Bi k, Bi v) => Bi (HM.HashMap k v) where
+  encode = encodeMapSkel HM.size $ \f acc ->
+      -- We need to encode the list with keys sorted in ascending order as
+      -- that's the only representation we accept during decoding.
+      foldr (uncurry f) acc . sortWith fst . HM.toList
+  decode = decodeMapSkel HM.fromList
+
+instance (Ord k, Bi k, Bi v) => Bi (Map k v) where
+  encode = encodeMapSkel M.size M.foldrWithKey
+  decode = decodeMapSkel M.fromAscList
+
+encodeSetSkel :: Bi a
+              => (s -> Int)
+              -> ((a -> E.Encoding -> E.Encoding) -> E.Encoding -> s -> E.Encoding)
+              -> s
+              -> E.Encoding
+encodeSetSkel size foldFunction =
+    encodeContainerSkel E.encodeListLen size foldFunction (\a b -> encode a <> b)
+{-# INLINE encodeSetSkel #-}
+
+decodeSetSkel :: Bi a => ([a] -> c) -> D.Decoder s c
+decodeSetSkel fromList = do
+  n <- D.decodeListLen
+  fmap fromList (replicateM n decode)
+{-# INLINE decodeSetSkel #-}
+
+instance (Hashable a, Ord a, Bi a) => Bi (HashSet a) where
+  encode = encodeSetSkel HS.size $ \f acc ->
+      -- We need to encode the list sorted in ascending order as that's the only
+      -- representation we accept during decoding.
+      foldr f acc . sort . HS.toList
+  decode = decodeSetSkel HS.fromList
+
+instance (Ord a, Bi a) => Bi (Set a) where
+  encode = encodeSetSkel S.size S.foldr
+  decode = decodeSetSkel S.fromAscList
```

Comment:
Don't worry -- as it have been pointed out for Maps by @neongreen , our `Bi` instances for `Set`, `Map` and `HashMap` do not check canonicity when they should. This will be all fixed today.